### ISSUE TITLE
Implicit cast between type and function pointer

### ISSUE THREAD

**Issue by sergeipetrashko8** [dotnet/roslyn#58718]
**Version Used**: Visual Studio 2022 | .NET 6.0 Console project

**Steps to Reproduce**:

Write the following code:

```cs
unsafe
{
    PointerDelegate<int, int> dp = (PointerDelegate<int, int>)(&Mtd);
}

static int Mtd(int arg) => arg;

public readonly struct PointerDelegate<T, TResult>
{
    private unsafe readonly delegate*<T,TResult> _pointer;

    public unsafe PointerDelegate(delegate*<T, TResult> pointer)
    {
        this._pointer = pointer;
    }

    public TResult Invoke(T param)
    {
        unsafe
        {
            return this._pointer(param);
        }
    }

    public unsafe static implicit operator PointerDelegate<T, TResult>(delegate*<T, TResult> pointer)
    {
        return new(pointer);
    }
}
```

The rule IDE0004 indicates that there is an unnecessary cast between the PointerDelegate<int, int> and the method group.

![image](https://user-images.githubusercontent.com/43956921/148594004-35a535e2-1b9b-4cee-bf97-b65b34b19145.png)

If the unnecessary cast is removed the following compiler error (CS8812) is appeared:

![image](https://user-images.githubusercontent.com/43956921/148594954-2e9b9f21-144c-49f6-8da1-07bdf27b41d8.png)

I believe there should not be any error using such implicit cast.

**Comment by Youssef1313** [comment]
@sergeipetrashko8 What version of Visual Studio are you on? The whole logic behind IDE0004 was completely rewritten and is shipped into the latest preview. Need to confirm whether this bug exists with the new logic or not.

**Comment by sergeipetrashko8** [comment]
@Youssef1313 I use 17.0.4 version. I don't have preview version installed.

**Comment by Youssef1313** [comment]
@sergeipetrashko8 I just tested quickly on the latest `main` and it seems to be working correctly. You should see this fixed in 17.1 when released.

Will let @CyrusNajmabadi weigh on whether a unit test should be added.

**Comment by CyrusNajmabadi** [comment]
Yes.  A unit test would be good here.

**Comment by Ziflin** [comment]
Can anyone confirm how this was fixed? I'm using Visual Studio v17.9.6 and still getting CS8812 when trying to use an implicit cast. An explicit cast does work, so do implicit casts just not work with function pointers?

**Comment by CyrusNajmabadi** [comment]
@Ziflin It sounds like you might be describing some other issue.  can you please open a new issue explaining what problem you're running into clearly?  Thanks!
